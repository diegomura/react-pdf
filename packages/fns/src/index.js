export { default as adjust } from './adjust';
export { default as asyncCompose } from './asyncCompose';
export { default as capitalize } from './capitalize';
export { default as castArray } from './castArray';
export { default as compose } from './compose';
export { default as dropLast } from './dropLast';
export { default as evolve } from './evolve';
export { default as get } from './get';
export { default as isNil } from './isNil';
export { default as last } from './last';
export { default as mapValues } from './mapValues';
export { default as matchPercent } from './matchPercent';
export { default as omit } from './omit';
export { default as pick } from './pick';
export { default as repeat } from './repeat';
export { default as reverse } from './reverse';
export { default as upperFirst } from './upperFirst';
