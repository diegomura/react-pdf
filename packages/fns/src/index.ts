export { default as adjust } from './adjust.js';
export { default as asyncCompose } from './asyncCompose.js';
export { default as capitalize } from './capitalize.js';
export { default as castArray } from './castArray.js';
export { default as compose } from './compose.js';
export { default as dropLast } from './dropLast.js';
export { default as evolve } from './evolve.js';
export { default as get } from './get.js';
export { default as isNil } from './isNil.js';
export { default as last } from './last.js';
export { default as mapValues } from './mapValues.js';
export { default as matchPercent } from './matchPercent.js';
export { default as omit } from './omit.js';
export { default as pick } from './pick.js';
export { default as repeat } from './repeat.js';
export { default as reverse } from './reverse.js';
export { default as upperFirst } from './upperFirst.js';
export { default as without } from './without.js';
export { default as parseFloat } from './parseFloat.js';
