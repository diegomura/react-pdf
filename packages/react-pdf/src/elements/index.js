import Text from './Text';
import View from './View';
import Page from './Page';
import Document from './Document';
import omit from 'lodash/fp/omit';

function createElement(type, props, root) {
  let instance;

  switch (type) {
    case 'ROOT':
      instance = new Document();
      break;
    case 'DOCUMENT':
      /*
       Since the Document instance is being created by the root element,
       all we do when the renderer gets to <Document /> is injecting
       it's props and return the root
      */
      root.setProps(omit('children', props));
      instance = root;
      break;
    case 'PAGE':
      instance = new Page(props, root);
      break;
    case 'TEXT':
      instance = new Text(props, root);
      break;
    case 'VIEW':
      instance = new View(props, root);
      break;
    default:
      instance = undefined;
  }

  return instance;
}

export { createElement };
