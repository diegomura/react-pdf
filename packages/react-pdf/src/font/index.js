import isUrl from 'is-url';
import standardFonts from './standard';
import { fetchFont } from '../utils/font';

let fonts = {};

const register = (src, { family, ...otherOptions }) => {
  fonts[family] = {
    src,
    loaded: false,
    data: null,
    ...otherOptions,
  };
};

const getRegisteredFonts = () => Object.keys(fonts);

const load = async (fontFamily, doc) => {
  const font = fonts[fontFamily];

  // We cache the font to avoid fetching it many time
  if (font && !font.data) {
    font.data = isUrl(font.src) ? await fetchFont(font.src) : font.src;
  }

  // If the font wasn't added to the document yet (aka. loaded), we do.
  // This prevents calling `registerFont` many times for the same font.
  // Fonts loaded state will be resetted after document is closed.
  if (font && !font.loaded) {
    font.loaded = true;
    doc.registerFont(fontFamily, font.data);
  }

  if (!font && !standardFonts.includes(fontFamily)) {
    throw new Error(
      `Font familiy not registered: ${fontFamily}. Please register it calling Font.register() method.`,
    );
  }
};

const reset = () => {
  for (const font in fonts) {
    if (fonts.hasOwnProperty(font)) {
      fonts[font].loaded = false;
    }
  }
};

const clear = () => {
  fonts = {};
};

export default {
  register,
  getRegisteredFonts,
  load,
  clear,
  reset,
};
