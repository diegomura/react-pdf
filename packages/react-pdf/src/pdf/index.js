/* global Blob */
import transformTree from './transformer';

const pdf = input => {
  const header = '%PDF-1.7\n';

  // Creates the footer of the PDF document
  function buildCrossReferenceTable(objects) {
    let offset = header.length - 1;

    const pointers = objects
      .map(object => {
        const pointer = `${('0000000000' + offset).slice(-10)} 00000 n`;
        offset += object.value.length;
        return pointer;
      })
      .join('\n');

    return 'xref\n' +
      `0 ${objects.length + 1}\n` +
      '0000000000 65535 f\n' +
      `${pointers}\n` +
      'trailer\n' +
      '<<\n' +
      `/Size ${objects.length}\n` +
      '/Root 1 0 R\n' +
      '>>' +
      'startxref\n' +
      `${offset}\n` +
      '%%EOF';
  }

  function buildDocument(input) {
    // Transforms the nodes trees into a more suitable data structure
    const objects = transformTree(input);

    // Builds the body and footer of the document
    const body = objects.map(object => object.value).join('\n\n');
    const footer = buildCrossReferenceTable(objects);

    return header + body + footer;
  }

  function toBlob() {
    return new Blob([buildDocument(input)], {
      type: 'application/pdf',
    });
  }

  function toBuffer() {
    return new Buffer(buildDocument(input));
  }

  return {
    toBuffer,
    toBlob,
  };
};

export default pdf;
