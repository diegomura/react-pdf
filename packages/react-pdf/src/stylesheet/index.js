import matchMedia from 'media-engine';
import transformStyles from './transformStyles';

const create = styles => styles;

const flatten = input => {
  if (!Array.isArray(input)) {
    input = [input];
  }

  const result = input.reduce((acc, style) => {
    if (style) {
      Object.keys(style).forEach(key => {
        if (style[key]) {
          acc[key] = style[key];
        }
      });
    }

    return acc;
  }, {});

  return result;
};

const resolveMediaQueries = (input, container) => {
  const result = Object.keys(input).reduce((acc, key) => {
    if (/@media/.test(key)) {
      return {
        ...acc,
        ...matchMedia({ [key]: input[key] }, container),
      };
    }

    return { ...acc, [key]: input[key] };
  }, {});

  return result;
};

const resolve = (styles, container) => {
  if (!styles) {
    return null;
  }

  styles = flatten(styles);
  styles = resolveMediaQueries(styles, container);

  return transformStyles(styles);
};

const absoluteFillObject = {
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

module.exports = {
  hairlineWidth: 1,
  create,
  resolve,
  flatten,
  absoluteFillObject,
};
