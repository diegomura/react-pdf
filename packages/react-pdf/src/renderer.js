'use strict';

import ReactFiberReconciler from 'react-reconciler';
import emptyObject from 'fbjs/lib/emptyObject';
import { createElement } from './elements';

const PDFRenderer = ReactFiberReconciler({
  appendInitialChild(parentInstance, child) {
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.document = child;
    }
  },

  createInstance(type, props, internalInstanceHandle) {
    return createElement(type, props, internalInstanceHandle);
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    return text;
  },

  finalizeInitialChildren(domElement, type, props) {
    return false;
  },

  getPublicInstance(instance) {
    return instance;
  },

  prepareForCommit() {
    // Noop
  },

  prepareUpdate(domElement, type, oldProps, newProps) {
    return true;
  },

  resetAfterCommit() {
    // Noop
  },

  resetTextContent(domElement) {
    // Noop
  },

  getRootHostContext() {
    return emptyObject;
  },

  getChildHostContext() {
    return emptyObject;
  },

  shouldSetTextContent(type, props) {
    return false;
  },

  now: () => {},

  useSyncScheduling: true,

  mutation: {
    appendChild(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.document = child;
      }
    },

    appendChildToContainer(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.document = child;
      }
    },

    insertBefore(parentInstance, child, beforeChild) {
      // noob
    },

    insertInContainerBefore(parentInstance, child, beforeChild) {
      // noob
    },

    removeChild(parentInstance, child) {
      parentInstance.removeChild(child);
    },

    removeChildFromContainer(parentInstance, child) {
      if (parentInstance.removeChild) {
        parentInstance.removeChild(child);
      }
    },

    commitTextUpdate(textInstance, oldText, newText) {
      textInstance = newText;
    },

    commitMount(instance, type, newProps) {
      // Noop
    },

    commitUpdate(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },
  },
});

export default PDFRenderer;
