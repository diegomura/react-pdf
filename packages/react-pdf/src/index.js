'use strict';

import { PDFRenderer, createElement } from './renderer';
import StyleSheet from './Stylesheet';
import pdf from './pdf/index.js';

/* Component constants */
const View = 'VIEW';
const Text = 'TEXT';
const Page = 'PAGE';
const Document = 'DOCUMENT';

export {
  PDFRenderer,
  View,
  Text,
  Page,
  Document,
  StyleSheet,
  createElement,
  pdf,
};
