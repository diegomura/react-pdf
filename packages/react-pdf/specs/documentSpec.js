import React from 'react';
import { Document } from '../src';
import MockDate from 'mockdate';
import render from './testRenderer';

describe('<Document />', () => {
  beforeEach(() => {
    MockDate.set(1434319925275);
  });

  const document = (props = {}) => <Document {...props} />;

  test('Should not render metadata if not specified', done => {
    render(document()).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Title metadata', done => {
    render(document({ title: 'Test' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Author metadata', done => {
    render(document({ author: 'John Doe' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Subject metadata', done => {
    render(document({ subject: 'Whatever' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Keywords metadata', done => {
    render(document({ keywords: 'lorem, ipsum' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Creator metadata', done => {
    render(document({ creator: 'other app' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });

  test('Should be able to set Producer metadata', done => {
    render(document({ producer: 'other app' })).then(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });
});
