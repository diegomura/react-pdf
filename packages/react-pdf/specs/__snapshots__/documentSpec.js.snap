// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Document /> Should be able to set Author metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
/Author (John Doe)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000181 00000 n 
0000000132 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
233
%%EOF
"
`;

exports[`<Document /> Should be able to set Creator metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (other app)
/CreationDate (D:20150614221205Z)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000162 00000 n 
0000000113 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
214
%%EOF
"
`;

exports[`<Document /> Should be able to set Keywords metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
/Keywords (lorem, ipsum)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000187 00000 n 
0000000138 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
239
%%EOF
"
`;

exports[`<Document /> Should be able to set Producer metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (other app)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000162 00000 n 
0000000113 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
214
%%EOF
"
`;

exports[`<Document /> Should be able to set Subject metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
/Subject (Whatever)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000182 00000 n 
0000000133 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
234
%%EOF
"
`;

exports[`<Document /> Should be able to set Title metadata 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
/Title (Test)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000176 00000 n 
0000000127 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
228
%%EOF
"
`;

exports[`<Document /> Should not render metadata if not specified 1`] = `
"%PDF-1.3
%����
3 0 obj
<<
/Producer (react-pdf)
/Creator (react-pdf)
/CreationDate (D:20150614221205Z)
>>
endobj
2 0 obj
<<
/Type /Catalog
/Pages 1 0 R
>>
endobj
1 0 obj
<<
/Type /Pages
/Count 0
/Kids []
>>
endobj
xref
0 4
0000000000 65535 f 
0000000162 00000 n 
0000000113 00000 n 
0000000015 00000 n 
trailer
<<
/Size 4
/Root 2 0 R
/Info 3 0 R
>>
startxref
214
%%EOF
"
`;
